File: src/main/java/no/skatteetaten/aurora/databasehotel/service/internal/DatabaseSchemaBuilder.java
package no.skatteetaten.aurora.databasehotel.service.internal;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import no.skatteetaten.aurora.databasehotel.dao.Schema;
import no.skatteetaten.aurora.databasehotel.dao.dto.Label;
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaData;
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaUser;
import no.skatteetaten.aurora.databasehotel.domain.DatabaseInstanceMetaInfo;
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema;
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchemaMetaData;
import no.skatteetaten.aurora.databasehotel.domain.User;
import no.skatteetaten.aurora.databasehotel.service.JdbcUrlBuilder;
import no.skatteetaten.aurora.databasehotel.service.SchemaSize;
import no.skatteetaten.aurora.databasehotel.utils.MapUtils;

public class DatabaseSchemaBuilder {

    private final DatabaseInstanceMetaInfo metaInfo;

    private final JdbcUrlBuilder jdbcUrlBuilder;

    public DatabaseSchemaBuilder(DatabaseInstanceMetaInfo metaInfo, JdbcUrlBuilder jdbcUrlBuilder) {

        this.metaInfo = metaInfo;
        this.jdbcUrlBuilder = jdbcUrlBuilder;
    }

    public Set<DatabaseSchema> createMany(
        List<SchemaData> schemaDataList,
        List<Schema> schemas,
        List<SchemaUser> users,
        List<Label> labels,
        List<SchemaSize> schemaSizes) {

        Map<String, SchemaData> schemaDataIndex = MapUtils.createUniqueIndex(schemaDataList, SchemaData::getName);
        Map<String, SchemaUser> userIndex = MapUtils.createUniqueIndex(users, SchemaUser::getSchemaId);
        Map<String, List<Label>> labelIndex = MapUtils.createNonUniqueIndex(labels, Label::getSchemaId);
        Map<String, SchemaSize> schemaSizeIndex = MapUtils.createUniqueIndex(schemaSizes, SchemaSize::getOwner);
        return schemas.stream()
            .filter(schema -> schemaDataIndex.containsKey(schema.getUsername()))
            .map(schema -> {
                SchemaData schemaData = schemaDataIndex.get(schema.getUsername());

                Optional<SchemaUser> user = MapUtils.maybeGet(userIndex, schemaData.getId());
                List<SchemaUser> schemaUsers = new ArrayList<>();
                user.ifPresent(schemaUsers::add);

                Optional<List<Label>> schemaLabels = MapUtils.maybeGet(labelIndex, schemaData.getId());
                SchemaSize schemaSize = schemaSizeIndex.get(schema.getUsername());

                return createOne(schemaData, schema, schemaUsers, schemaLabels, schemaSize);
            })
            .collect(Collectors.toSet());
    }

    public DatabaseSchema createOne(SchemaData schemaData, Schema schema, List<SchemaUser> users,
        Optional<List<Label>> labelsOption, SchemaSize schemaSize) {

        return createOne(schemaData, schema, users, labelsOption, schemaSize, DatabaseSchema.Type.MANAGED);
    }

    public DatabaseSchema createOne(SchemaData schemaData, Schema schema, List<SchemaUser> users,
        List<Label> labelsOption, SchemaSize schemaSize, DatabaseSchema.Type type) {
        return createOne(schemaData, schema, users, Optional.ofNullable(labelsOption), schemaSize, type);
    }

    public DatabaseSchema createOne(SchemaData schemaData, Schema schema, List<SchemaUser> users,
        Optional<List<Label>> labelsOption, SchemaSize schemaSize, DatabaseSchema.Type type) {

        String jdbcUrl = jdbcUrlBuilder.create(metaInfo.getHost(), metaInfo.getPort(), schema.getUsername());

        DatabaseSchema databaseSchema = new DatabaseSchema(
            schemaData.getId(),
            metaInfo,
            jdbcUrl,
            schema.getUsername(),
            schema.getCreated(),
            schema.getLastLogin(),
            createMetaData(schemaSize),
            type
        );
        users.forEach(u -> databaseSchema.addUser(new User(u.getId(), u.getUsername(), u.getPassword(), u.getType())));

        Map<String, String> labelMap = new HashMap<>();
        labelsOption.ifPresent(labels -> labels.forEach(label -> labelMap.put(label.getName(), label.getValue())));
        databaseSchema.setLabels(labelMap);

        return databaseSchema;
    }

    private DatabaseSchemaMetaData createMetaData(SchemaSize schemaSize) {

        double size = Optional.ofNullable(schemaSize)
            .map(SchemaSize::getSchemaSizeMb)
            .map(BigDecimal::doubleValue)
            .orElse(0.0);
        return new DatabaseSchemaMetaData(size);
    }
}

File: src/main/kotlin/no/skatteetaten/aurora/databasehotel/service/DatabaseSchemaBuilder.kt
package no.skatteetaten.aurora.databasehotel.service

import java.math.BigDecimal
import no.skatteetaten.aurora.databasehotel.dao.Schema
import no.skatteetaten.aurora.databasehotel.dao.dto.Label
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaData
import no.skatteetaten.aurora.databasehotel.dao.dto.SchemaUser
import no.skatteetaten.aurora.databasehotel.domain.DatabaseInstanceMetaInfo
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema.Type
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchema.Type.MANAGED
import no.skatteetaten.aurora.databasehotel.domain.DatabaseSchemaMetaData
import no.skatteetaten.aurora.databasehotel.domain.User

data class DatabaseSchemaBuilder(
    private val metaInfo: DatabaseInstanceMetaInfo,
    private val jdbcUrlBuilder: JdbcUrlBuilder,
    val users: List<SchemaUser>,
    val schemas: List<Schema>,
    val labels: List<Label>,
    val schemaSizes: List<SchemaSize>
) {
    private val schemaIndex = schemas.map { it.username to it }.toMap()
    private val userIndex = users.map { it.schemaId!! to it }.toMap()
    private val labelIndex = labels.groupBy { it.schemaId!! }
    private val schemaSizeIndex = schemaSizes.map { it.owner to it }.toMap()

    fun createMany(schemaDataList: List<SchemaData>, type: Type = MANAGED) = schemaDataList
        .filter { schemaIndex.containsKey(it.name) }
        .map { createOne(it, type) }.toSet()

    fun createOne(schemaData: SchemaData, type: Type = MANAGED): DatabaseSchema {

        val schema = schemaIndex[schemaData.name]
            ?: error("Missing SchemaData for Schema ${schemaData.name}")

        val schemaUsers = listOfNotNull(userIndex[schemaData.id])
        val schemaLabels = labelIndex[schemaData.id] ?: emptyList()
        val schemaSize = schemaSizeIndex[schema.username]

        return createOne(schemaData, schema, schemaUsers, schemaLabels, schemaSize, type)
    }

    private fun createOne(
        schemaData: SchemaData,
        schema: Schema,
        users: List<SchemaUser>,
        labels: List<Label>,
        schemaSize: SchemaSize?,
        type: Type = MANAGED
    ): DatabaseSchema {

        val jdbcUrl = jdbcUrlBuilder.create(metaInfo.host, metaInfo.port, schema.username)

        val databaseSchema = DatabaseSchema(
            schemaData.id,
            schemaData.active,
            metaInfo,
            jdbcUrl,
            schema.username,
            schemaData.createdDate,
            schema.lastLogin,
            schemaData.setToCooldownAt,
            schemaData.deleteAfter,
            createMetaData(schemaSize = schemaSize ?: SchemaSize(schema.username, BigDecimal.ZERO)),
            type
        )
        users.forEach { databaseSchema.addUser(User(it.id!!, it.username!!, it.password!!, it.type!!)) }
        databaseSchema.labels = labels.map { it.name!! to it.value }.toMap()

        return databaseSchema
    }

    companion object {

        @JvmStatic
        private fun createMetaData(schemaSize: SchemaSize?): DatabaseSchemaMetaData =
            DatabaseSchemaMetaData(schemaSize?.schemaSizeMb?.toDouble() ?: 0.0)
    }
}
